libs/libc/tls/tls_free.c

```diff
@@ -59,26 +56,22 @@
 
 int tls_free(int tlsindex)
 {
-  FAR struct tcb_s *rtcb = this_task();
-  FAR struct task_group_s *group = rtcb->group;
+  FAR struct task_info_s *tinfo = task_get_info();
   tls_ndxset_t mask;
-  irqstate_t flags;
   int ret = -EINVAL;
 
-  DEBUGASSERT((unsigned)tlsindex < CONFIG_TLS_NELEM && group != NULL);
+  DEBUGASSERT((unsigned)tlsindex < CONFIG_TLS_NELEM && tinfo != NULL);
   if ((unsigned)tlsindex < CONFIG_TLS_NELEM)
     {
       /* This is done in a critical section here to avoid concurrent
        * modification of the group TLS index set.
        */
 
       mask  = (1 << tlsindex);
-      flags = spin_lock_irqsave(NULL);
-
-      DEBUGASSERT((group->tg_tlsset & mask) != 0);
-      group->tg_tlsset &= ~mask;
-      spin_unlock_irqrestore(NULL, flags);
-
+      sem_wait(&tinfo->ta_tlssem);
```

Comment:
Should _SEM_WAIT be used?  It should be if this could ever be called from within the OS.  If not, then it should not be built if __KERNEL__ is defined.